contexts/did-v1.jsonld

```diff
@@ -116,6 +116,10 @@
     },
     "publicKeyBase58": "sec:publicKeyBase58",
     "publicKeyPem": "sec:publicKeyPem",
+    "publicKeyJwk": {
+      "@id": "sec:publicKeyJwk",
+      "@type": "@json"
```

Comment:
Are we aware of any issues with JCS and RDF Dataset Normalization wrt. JWKs? I'd like someone to do a thorough security review on that if we're going to use `@json` here. That goes for any time we use `@json` ... I think this might be the first ever use of that... and will require another feature of JSON-LD 1.1 processors. The other alternative is to define all items in JOSE registry.

We really shouldn't merge until someone has done a proper review of `@json` interplay w/ RDF Dataset Normalization since this stuff is going to end up in DID Documents.

Reply:
I can report that we've been using `@json` literals with success in conjunction with RDF Dataset Normalization for several months now. No issues to report.

Reply:
We have also been using them... they are extremly valuable when handling embedded JSON Schemas in VCs and when handling data models that keep changing based on customer / stakeholder feedback... our latest interop tests with certified mill test reports use them, as well as our previous work with Workday on vc-json-schemas

Reply:
@msporny there are also CI tests which confirm that this working correctly... thats part of why we added the CI system.

Reply:
> also CI tests which confirm that this working correctly

My default with any sort of security related PR is to be conservative and confirm what people are asserting before agreeing to the PR. :)

For example, I expect that the CI system is using the same library to normalize, sign, and verify. Which means, there is only one implementation, which means that if that implementation has a bug w/ JCS that we won't catch it.

Another example: Do we need anything in the security considerations section related to publicKeyJwk - since it's `@json`, that enables people to put literally anything in there... like key_ops (what happens when this conflicts w/ the verification method?), x5t (so, we allow SHA-1 fingerprints when we know that's a bad idea?), and x5c (wait a sec, we allow certificate chains!?).

Are we subsetting? If not, why not? See my point? I'm questioning whether or not there has been a proper security review on the use of `@json` with `publicKeyJwk`. If there has been, I'd like to see it. To be clear, I'm a +1 for including publicKeyJwk... but a -1 for handing developers a foot gun, which blindly combining `publicKeyJwk` and `@json` does.

Reply:
https://www.ietf.org/rfc/rfc7517.txt

> Additional members can be present in the JWK; if not understood by
>  implementations encountering them, they MUST be ignored.  Member
>  names used for representing key parameters for different keys types
>  need not be distinct.  Any new member name should either be
>  registered in the IANA "JSON Web Key Parameters" registry established
>  by Section 8.1 or be a value that contains a Collision-Resistant
>  Name.

Reply:
Sounds like this WG is attempting to redefine something which is has no authority to.

Reply:
It's common for applications of standards to be able to subset the standards that they use to restrict the security attack surface.

Reply:
I agree, I heard this argument loud and clear from the "Pure JSON" camp... its a great reason not to include dependencies on w3id.org vocabularies or JSON-LD in its entirety. I could make the same argument about CBOR.

So to be clear, you are suggesting that we make changes to restrict `publicKeyJwk` beyond what the RFC states, including disallowing features such as `key_ops`,  `x5c`, which might be used by JSON only did methods, in lieu of the vocabulary that is currently supposed to be defined by w3id.org?

Here are a list of properties that were added to did core context for which no definition exists... 

https://github.com/w3c/did-core-registries/pull/21

Let's start by gaining consensus that we are not going to allow undefined properties in did core. 

Next for each of the properties that we will be adding, we can of course define a JSON Schema that locks the property down further (eliminating competing features of existing systems, such as `key_ops`)... I expect that will be met with pretty strong objection (at least from me).

If you are saying we won't get `publicKeyJwk` in did core until we restrict it, then we should apply the same level of rigor to the existing properties... right away, and then work towards a tractable solution.

What I am attempting to do here, is support JOSE and the existing ecosystem, while also expanding support for JSON-LD.

I'm not assuming malice on your part, I believe that your position is that support for `key_ops`, or the ability for controllers to add properties to their JWK, based on the RFC, are security concerns, which we should not rush to enable. 

I want to understand how we will be addressing these concerns, what can I, as the PR author, do to address your concerns?

Reply:
> So to be clear, you are suggesting that we make changes to restrict `publicKeyJwk` beyond what the RFC states

Yes, I'm saying that we should discuss doing that... because if we don't, we're going to have to define, in the registry entry, all the ways you deal with verification methods stating one thing and then JWKs stating something else. To provide an example, we just had to define text in the DID Core spec around how to reconcile `verificationMethodX.id` and `verificationMethodX.publicKeyJwk.kid`. I expect that we might need to do the same for some entries in the JOSE JWK registry, and I was asking if anyone has done this work yet. 

> including disallowing features such as `key_ops`, `x5c`, which might be used by JSON only did methods, in lieu of the vocabulary that is currently supposed to be defined by w3id.org?

I'm saying that it's starting to become clear that nobody has talked about this, so we should talk about it.

> Here are a list of properties that were added to did core context for which no definition exists...
> 
> #21

Hmmm... that PR seems to be proposing breaking the entire DID ecosystem that utilizes JSON-LD.

> Let's start by gaining consensus that we are not going to allow undefined properties in did core.

I don't think having that discussion is going to help this issue. That discussion has it's own set of arguments... some folks want items to be ignored, others want them to throw errors. We'll further delay this item by discussing that item.

> Next for each of the properties that we will be adding, we can of course define a JSON Schema that locks the property down further (eliminating competing features of existing systems, such as `key_ops`)... I expect that will be met with pretty strong objection (at least from me).

The spec will either have to define what happens when `key_ops` differs from what the verification method does, or we'll have to disallow it, or maybe someone else will come up with a bright idea when we discuss this. My point is that it's clear that this hasn't been discussed yet.

> If you are saying we won't get `publicKeyJwk` in did core until we restrict it, then we should apply the same level of rigor to the existing properties... right away, and then work towards a tractable solution.

Yes, we should apply the same level of rigor to everything in the spec... but if we do it right away, we'll grind spec development to a halt, so I think it would have the opposite effect of what you're going for.

> What I am attempting to do here, is support JOSE and the existing ecosystem, while also expanding support for JSON-LD.

Sure, I get that... and that's good. We also need to be careful about the security attack surface we're opening up.

> I'm not assuming malice on your part, I believe that your position is that support for `key_ops`, or the ability for controllers to add properties to their JWK, based on the RFC, are security concerns, which we should not rush to enable.

Good, because I'm not acting out of malice... I'm doing my job as an Editor, which entails making sure that I understand what's going into the spec and to make sure it has consensus. I don't think we've had these set of discussions yet.

> I want to understand how we will be addressing these concerns, what can I, as the PR author, do to address your concerns?

To move this along quickly, we probably need to raise an issue on the DID Core spec noting that we haven't done a proper security review on the way that verification methods and publicKeyJwk interact (all of the possibilities of what you can put in publicKeyJwk). We should point that issue out via a `<p class="issue">...</p>` note in the spec. At that point, this PR could probably go in and then we can process the issue in did-core.

Reply:
could you define a JSON-LD context for the JWK data model and reference that with the type being JWK or are contexts not allowed at this level?

Reply:
You can embed such a definition... but it would NEVER be RFC compliant if it did not allow arbitrary json, because the RFC allows arbitrary json... and some folks might rely on the RFC definition of what a JWK is.

I have raised an issue in did core, per feedback: https://github.com/w3c/did-core/issues/240